ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -31,215 +81,264 @@ where
+            prover::{TwoNPlusOne, UVPolynomial, ZeroKnowledgeProof},
+        },
+        secret_sharing::SharedValue,
+    };
+
+    impl Default for Fp31 {
+        fn default() -> Self {
+            Self::ZERO
         }
-        ZeroKnowledgeProof::new(proof)
     }
 
-    pub fn compute_final_proof<λ, I, J>(
-        u: I,
-        v: J,
+    // todo: deprecate
+    fn compute_final_proof<F, R>(
+        uv: &UVPolynomial<F, R>,
         p_0: F,
         q_0: F,
-        lagrange_table: &LagrangeTable<F, Sum<λ, U1>, λ>,
-    ) -> ZeroKnowledgeProof<F, TwoNPlusOne<λ>>
+        lagrange_table: &LagrangeTable<F, Sum<R, U1>, R>,
+    ) -> GenericArray<F, TwoNPlusOne<R>>
     where
-        λ: ArrayLength + Add + Add<U1>,
-        <λ as Add>::Output: Add<U1>,
-        <<λ as Add>::Output as Add<U1>>::Output: ArrayLength,
-        <λ as Add<U1>>::Output: ArrayLength,
-        I: IntoIterator<Item = F>,
-        J: IntoIterator<Item = F>,
-        I::IntoIter: ExactSizeIterator,
-        J::IntoIter: ExactSizeIterator,
+        F: PrimeField,
+        R: Add + Add<U1> + ArrayLength,
+        <R as Add>::Output: Add<U1>,
+        <<R as Add>::Output as Add<U1>>::Output: ArrayLength,
+        <R as Add<U1>>::Output: ArrayLength,
     {
-        let mut u = u.into_iter();
-        let mut v = v.into_iter();
-
-        assert_eq!(u.len(), λ::USIZE); // We should pad with zeroes eventually
-        assert_eq!(v.len(), λ::USIZE); // We should pad with zeroes eventually
-
-        let mut p = GenericArray::<F, Sum<λ, U1>>::generate(|_| F::ZERO);
-        let mut q = GenericArray::<F, Sum<λ, U1>>::generate(|_| F::ZERO);
-        let mut proof: GenericArray<F, TwoNPlusOne<λ>> = GenericArray::generate(|_| F::ZERO);
+        let mut p = GenericArray::<F, Sum<R, U1>>::generate(|_| F::ZERO);
+        let mut q = GenericArray::<F, Sum<R, U1>>::generate(|_| F::ZERO);
+        let mut proof: GenericArray<F, TwoNPlusOne<R>> = GenericArray::generate(|_| F::ZERO);
         p[0] = p_0;
         q[0] = q_0;
         proof[0] = p_0 * q_0;
 
-        for i in 0..λ::USIZE {
-            let x = u.next().unwrap_or(F::ZERO);
-            let y = v.next().unwrap_or(F::ZERO);
-            p[i + 1] = x;
-            q[i + 1] = y;
-            proof[i + 1] += x * y;
+        for i in 0..R::USIZE {
+            p[i + 1] = uv.0[i];
+            q[i + 1] = uv.1[i];
+            proof[i + 1] += uv.0[i] * uv.1[i];
         }
         // We need a table of size `λ + 1` since we add a random point at x=0
         let p_extrapolated = lagrange_table.eval(&p);
         let q_extrapolated = lagrange_table.eval(&q);
 
         for (i, (x, y)) in zip(p_extrapolated.into_iter(), q_extrapolated.into_iter()).enumerate() {
-            proof[λ::USIZE + 1 + i] += x * y;
+            proof[R::USIZE + 1 + i] += x * y;
         }
 
-        ZeroKnowledgeProof::new(proof)
+        proof
     }
 
-    pub fn gen_challenge_and_recurse<λ, I, J>(
```

Comment:
It's very hard to review this PR because the corresponding function is not displayed nearby in the diff view.

This would not be a problem if you had cut this PR into a smaller number of more incremental changes, so that one could easily see the side-by-side comparison of how this method looked before and after.

Reply:
I think this is independent of the more complex parts of this PR in proof_generation.rs. 

I think it is a result of that I have made many iterative changes to prover.rs and finally decided to use different structs and associate gen_challenge_and_recurse and compute_proof with different structs.